File: codechef/JCWR2019/c.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long
const ll mod = 1e9 + 7;
int T, n, odd, even, other;
ll qmod(ll base, ll n) {
	ll res = 1;
	while (n) {
		if (n & 1) res = res * base % mod;
		base = base * base % mod;
		n >>= 1;
	}
	return res;
}

int main() {
	scanf("%d", &T);
	while (T--) {
		scanf("%d", &n);
		odd = 0; even = 0; other = 0;
		for (int i = 1, x, y; i <= n; ++i) {
			scanf("%d%d", &x, &y);
			if (x % 2 != y % 2) ++other;
			else if (x % 2 == 1) ++odd;
			else ++even;
		}		
		if (other > 0) printf("%lld\n", qmod(2, n - 1));
		else if (odd % 2 == 0) puts("0");
		else printf("%lld\n", qmod(2, n));
	}
	return 0;
}

File: codechef/JCWR2019/c.py
T = int(input())
for i in range(0, T):
    n = int(input())
    odd = 0
    even = 0
    other = 0
    for j in range(0, n):
        x, y = map(int, input().split())
        if x % 2 != y % 2:
            other += 1
        elif x % 2 == 1 and y % 2 == 1:
            odd += 1
        else:
            even += 1
    if other > 0:
        print(2 ** (n - 1))
    elif odd % 2 == 0:
        print(0)
    else:
        print(2 ** n)
